Mathlib/Tactic/Ring/Compare.lean

```diff
@@ -0,0 +1,239 @@
+  /- For a numeral `0 ≤ c`, `x ≤ c + x` -/
+  | _, .add (.const (e := xb) cb hypb) vb' => do
+    unless va.eq vb' do return .error notComparable
+    let rxb := NormNum.Result.ofRawRat cb xb hypb
+    let NormNum.Result.isTrue pf ← NormNum.evalLE.core lα rz rxb | return .error tooSmall
+    pure <| .ok (q(le_add_of_nonneg_left (a := $a) $pf):)
+  | _, _ => return .error notComparable
+
+/-- In a commutative semiring, given `Ring.ExSum` objects `va`, `vb` which differ by a positive
+constant, construct a proof of `$a < $b`, where `a` (resp. `b`) is the expression in the semiring to
+which `va` (resp. `vb`) evaluates. -/
+def evalLT {v : Level} {α : Q(Type v)} (_ : Q(StrictOrderedCommSemiring $α)) {a b : Q($α)}
+    (va : Ring.ExSum q(tc₄ $α) a) (vb : Ring.ExSum q(tc₄ $α) b) :
+    MetaM (Except ExceptType Q($a < $b)) := do
+  let lα : Q(LT $α) := q(tc₆ $α)
+  assumeInstancesCommute
+  let ⟨_, pz⟩ ← NormNum.mkOfNat α q(tc₅ $α) (mkRawNatLit 0)
+  let rz : NormNum.Result q((0:$α)) :=
+    NormNum.Result.isNat q(tc₅ $α) (mkRawNatLit 0) (q(NormNum.isNat_ofNat $α $pz):)
+  match va, vb with
+  /- `0 < 0` -/
+  | .zero, .zero => return .error tooSmall
+  /- For numerals `c₁` and `c₂`, `c₁ + x < c₂ + x` if `c₁ < c₂` -/
+  | .add (b := a') (.const (e := xa) ca hypa) va', .add (.const (e := xb) cb hypb) vb' => do
+    unless va'.eq vb' do return .error notComparable
+    let rxa := NormNum.Result.ofRawRat ca xa hypa
+    let rxb := NormNum.Result.ofRawRat cb xb hypb
+    let NormNum.Result.isTrue pf ← NormNum.evalLT.core lα rxa rxb | return .error tooSmall
+    pure <| .ok (q(add_lt_add_right $pf $a'):)
+  /- For a numeral `c < 0`, `c + x < x` -/
+  | .add (.const (e := xa) ca hypa) va', _ => do
+    unless va'.eq vb do return .error notComparable
+    let rxa := NormNum.Result.ofRawRat ca xa hypa
+    let NormNum.Result.isTrue pf ← NormNum.evalLT.core lα rxa rz | return .error tooSmall
+    have pf : Q($xa < 0) := pf
+    pure <| .ok (q(add_lt_of_neg_left $b $pf):)
+  /- For a numeral `0 < c`, `x < c + x` -/
+  | _, .add (.const (e := xb) cb hypb) vb' => do
+    unless va.eq vb' do return .error notComparable
+    let rxb := NormNum.Result.ofRawRat cb xb hypb
+    let NormNum.Result.isTrue pf ← NormNum.evalLT.core lα rz rxb | return .error tooSmall
+    pure <| .ok (q(lt_add_of_pos_left $a $pf):)
+  | _, _ => return .error notComparable
+
+theorem le_congr {α : Type*} [LE α] {a b c d : α} (h1 : a = b) (h2 : b ≤ c) (h3 : d = c) :
+    a ≤ d := by
+  rwa [h1, h3]
+
+theorem lt_congr {α : Type*} [LT α] {a b c d : α} (h1 : a = b) (h2 : b < c) (h3 : d = c) :
+    a < d := by
+  rwa [h1, h3]
+
+/-- Prove goals of the form `A ≤ B` in a commutative semiring, if the ring-normal forms of `A` and
+`B` differ by a nonnegative constant. -/
+def proveLE (g : MVarId) : MetaM Unit := do
+  let some (α, e₁, e₂) := (← whnfR <|← instantiateMVars <|← g.getType).le?
+    | throwError "ring failed: not of the form `A ≤ B`"
```

Comment:
Wouldn't it be better to keep this in `MetaM (Except ...)` so that you don't have to guess which tactic it will be called in? I find it a bit dangerous to the user experience to have tactic names hardcoded in non-user-facing code